autorest/adal/popkey.go

```diff
@@ -0,0 +1,152 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+//PoPKey - generic interface for PoP key properties and methods
+type PoPKey interface {
+	// encryption/signature algo
+	Alg() string
+	// kid
+	KeyID() string
+	// jwk that can be embedded in JWT w/ PoP token's cnf claim
+	JWK() string
+	// https://tools.ietf.org/html/rfc7638 compliant jwk thumbprint
+	JWKThumbprint() string
+	// req_cnf claim that can be included in access token request to AAD
+	ReqCnf() string
+	// sign payload using private key
+	Sign([]byte) ([]byte, error)
+}
+
+// software based pop key implementation of PoPKey
+type swKey struct {
+	key    *rsa.PrivateKey
+	keyID  string
+	jwk    string
+	jwkTP  string
+	reqCnf string
+}
+
+func (swk *swKey) Alg() string {
+	return "RS256"
+}
+
+func (swk *swKey) KeyID() string {
+	return swk.keyID
+}
+
+func (swk *swKey) JWK() string {
+	return swk.jwk
+}
+
+func (swk *swKey) JWKThumbprint() string {
+	return swk.jwkTP
+}
+
+func (swk *swKey) ReqCnf() string {
+	return swk.reqCnf
+}
+
+func (swk *swKey) Sign(payload []byte) ([]byte, error) {
+	return swk.key.Sign(rand.Reader, payload, crypto.SHA256)
+}
+
+func (swk *swKey) init(key *rsa.PrivateKey) {
+	swk.key = key
+
+	pubKey := &swk.key.PublicKey
+	e := big.NewInt(int64(pubKey.E))
+	eB64 := base64.RawURLEncoding.EncodeToString(e.Bytes())
+	n := pubKey.N
+	nB64 := base64.RawURLEncoding.EncodeToString(n.Bytes())
+
+	// compute JWK thumbprint
+	//jwk format - e, kty, n - in lexicographic order
+	// - https://tools.ietf.org/html/rfc7638#section-3.3
+	// - https://tools.ietf.org/html/rfc7638#section-3.1
+	jwk := fmt.Sprintf(`{"e":"%s","kty":"RSA","n":"%s"}`, eB64, nB64)
+	jwkS256 := sha256.Sum256([]byte(jwk))
+	swk.jwkTP = base64.RawURLEncoding.EncodeToString(jwkS256[:])
+
+	//req_cnf - base64URL("{"kid":"jwkTP","xms_ksl":"sw"}")
+	reqCnfJSON := fmt.Sprintf(`{"kid":"%s","xms_ksl":"sw"}`, swk.jwkTP)
+	swk.reqCnf = base64.RawURLEncoding.EncodeToString([]byte(reqCnfJSON))
+
+	//set keyID to jwkTP
+	swk.keyID = swk.jwkTP
+
+	//compute JWK to be included in JWT w/ PoP token's cnf claim
+	// - https://tools.ietf.org/html/rfc7800#section-3.2
+	swk.jwk = fmt.Sprintf(`{"e":"%s","kty":"RSA","n":"%s","alg":"RS256","kid":"%s"}`, eB64, nB64, swk.keyID)
+}
+
+func generateSwKey() (*swKey, error) {
+	swk := &swKey{}
+	// 1024 good enough for rsa popkey that would be rotated frequently ?
+	key, err := rsa.GenerateKey(rand.Reader, 1024)
+	if err != nil {
+		return nil, err
+	}
+
+	swk.init(key)
+	return swk, nil
+}
+
+var pswKey *swKey
```

Comment:
do these need to be globals? Could they be held on a type that encapsulates this logic?

Reply:
We could put these 2 inside a struct but would need a global instance of that struct, that requires initialization, perhaps in an init(). Found it an overkill for the need here. Is there a better option ?

Reply:
I see, if it has to be a global anyways, not a lot of value add. I was hoping something else could own it, but I'm not seeing much.